shell/platform/windows/compositor_opengl_unittests.cc

```diff
@@ -0,0 +1,192 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <memory>
+#include <vector>
+
+#include "flutter/impeller/renderer/backend/gles/gles.h"
+#include "flutter/shell/platform/windows/angle_surface_manager.h"
+#include "flutter/shell/platform/windows/compositor_opengl.h"
+#include "flutter/shell/platform/windows/flutter_windows_view.h"
+#include "flutter/shell/platform/windows/testing/engine_modifier.h"
+#include "flutter/shell/platform/windows/testing/flutter_windows_engine_builder.h"
+#include "flutter/shell/platform/windows/testing/mock_window_binding_handler.h"
+#include "flutter/shell/platform/windows/testing/windows_test.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace flutter {
+namespace testing {
+
+namespace {
+using ::testing::Return;
+
+const unsigned char* MockGetString(GLenum name) {
+  switch (name) {
+    case GL_VERSION:
+    case GL_SHADING_LANGUAGE_VERSION:
+      return (unsigned char*)"3.0";
+    default:
+      return (unsigned char*)"";
+  }
+}
+
+void MockGetIntegerv(GLenum name, int* value) {
+  *value = 0;
+}
+
+GLenum MockGetError() {
+  return GL_NO_ERROR;
+}
+
+void DoNothing() {}
+
+const impeller::ProcTableGLES::Resolver kMockResolver = [](const char* name) {
+  if (strcmp(name, "glGetString") == 0) {
+    return reinterpret_cast<void*>(&MockGetString);
+  } else if (strcmp(name, "glGetIntegerv") == 0) {
+    return reinterpret_cast<void*>(&MockGetIntegerv);
+  } else if (strcmp(name, "glGetError") == 0) {
+    return reinterpret_cast<void*>(&MockGetError);
+  } else {
+    return reinterpret_cast<void*>(&DoNothing);
+  }
+};
+
+class MockAngleSurfaceManager : public AngleSurfaceManager {
+ public:
+  MockAngleSurfaceManager() : AngleSurfaceManager(false) {}
+
+  MOCK_METHOD(bool, MakeCurrent, (), (override));
+
+ private:
+  FML_DISALLOW_COPY_AND_ASSIGN(MockAngleSurfaceManager);
+};
+
+class MockFlutterWindowsView : public FlutterWindowsView {
+ public:
+  MockFlutterWindowsView(std::unique_ptr<WindowBindingHandler> window)
+      : FlutterWindowsView(std::move(window)) {}
+  virtual ~MockFlutterWindowsView() = default;
+
+  MOCK_METHOD(bool, SwapBuffers, (), (override));
+
+ private:
+  FML_DISALLOW_COPY_AND_ASSIGN(MockFlutterWindowsView);
+};
+
+class CompositorOpenGLTest : public WindowsTest {
+ public:
+  CompositorOpenGLTest() = default;
+  virtual ~CompositorOpenGLTest() = default;
+
+ protected:
+  FlutterWindowsEngine* engine() { return engine_.get(); }
+  MockFlutterWindowsView* view() { return view_.get(); }
+  MockAngleSurfaceManager* surface_manager() { return surface_manager_; }
+
+  void UseHeadlessEngine() {
+    auto surface_manager = std::make_unique<MockAngleSurfaceManager>();
+    surface_manager_ = surface_manager.get();
+
+    FlutterWindowsEngineBuilder builder{GetContext()};
+
+    engine_ = builder.Build();
+    EngineModifier modifier(engine_.get());
+    modifier.SetSurfaceManager(surface_manager.release());
```

Comment:
As the reader, my first question would be "now that we're no longer the owner of `surface_manager`, is `EngineModifier` definitely freeing it? The answer is of course yes, but it may be worth making this more obvious at the call site. I can think of two ways to do that:

1. Add an `EngineModifier::SetSurfaceManager(std::unique_ptr<AngleSurfaceManager>)` override and `std::move(surface_manager)` here.
2. Add a comment `// Hand ownership of surface_manager to EngineModifier.`

Reply:
This is an excellent idea. I've opened https://github.com/flutter/engine/pull/48953 to clean this up. Once that lands I'll rebase and update this pull request.

Reply:
Fixed!